onnxruntime/core/providers/cpu/tensor/scatter.cc

```diff
@@ -32,16 +45,29 @@ ONNX_CPU_OPERATOR_KERNEL(
 
 template <class Tin, class Tdata>
 Status CopyScatterData(const Tensor* data_input, const Tensor* indices_input, const Tensor* updates_input,
-                       const int64_t axis, Tensor* data_output) {
+                       const int64_t axis, Tensor* data_output, bool is_scatter_elements) {
   const TensorShape& input_data_shape = data_input->Shape();
-  const Tin* indices_data = indices_input->template Data<Tin>();
+  const Tin* indices_data_raw= indices_input->template Data<Tin>();
   const auto num_indices = indices_input->Shape().Size();
+
+  std::vector<Tin> indices_data;
+  indices_data.reserve(num_indices);
+
+  auto axis_dim_limit = input_data_shape[axis];
+
   for (int64_t i = 0; i < num_indices; ++i) {
-    Tin idx = indices_data[i];
-    if (idx < 0 || idx >= input_data_shape[axis]) {
+    Tin idx = indices_data_raw[i];
+
+    if ((idx >= axis_dim_limit) ||  
+        // 'Scatter' does not support negative indices
+        (idx < 0 && !is_scatter_elements) ||
+        // 'ScatterElements' supports valid negative indices in range [-axis_dim_limit, -1]
+        (idx < -axis_dim_limit)) {
```

Comment:
https://github.com/onnx/onnx/pull/2260 clarifies negative index support for ScatterElements